### ISSUE TITLE
Normalize nested fields

### ISSUE THREAD

**Issue by joshgagnon** [erikras/redux-form#516]
Trying to combine nested fields with normalization, but can't see the intended way to do this.

I would have thought it would similar to:
```
const form = formReducer.normalize({
    'form': {
        'parent.child': someNormalizeFunction
    }
})...
```

Any suggestions?  Do I need to implement this?

**Comment by erikras** [comment]
I've never done that. I'll investigate and get back to you. :mag:

**Comment by erikras** [comment]
Yeah. The answer is that normalization of deep fields is currently not supported. :disappointed:

**Comment by Misaka42** [comment]
What about normalizing forms with formKey?
It seems that normalizer it attempting to set a global field on form with no regard to formKey.

**Comment by erikras** [comment]
@Misaka42 Normalizing with `formKey` is supposed to work. Someone PR'd a test for it, at least. :smile: As long as your subform is flat, I'd expect it to work.

**Comment by joshgagnon** [comment]
Normalizing with formKeys works fine for me (warning: I submitted the PR).  Hopefully I find some time to implement nested normalization.  

If I do have a stab at it, would you prefer style in the example given in the parent comment?

**Comment by Misaka42** [comment]
Accidentally submitted the issue anyway, and I can't close it for some reason.
Well, my form is not flat, but should that really matter?
If I, say, instead of trying normalizing a nested FIELD, normalize the whole nested object as I suggested previously, would that work?
Currently though, redux form just creates and empty field in form's root, as in the same level as form keys. Of course it's empty and of course it's not being normalizing.

**Comment by erikras** [comment]
@joshgagnon Indeed you are! My initial instinct is that for normalizing, for example, a phone number in:
```js
{
  billing: {
    phone: '555-555-5555'
  },
  shipping: {
    phone: '555-555-5555'
  }
}
```
It would be nice to nest the function:
```js
reducer.normalize({
  contactForm: {
    billing: {
       phone: normalizePhone
    },
    shipping: {
       phone: normalizePhone
    }
  }
})
```
But that won't let you do array values, e.g. `address.phones[]`, so maybe the string key is better. i.e.
```js
reducer.normalize({
  contactForm: {
    'billing.phone': normalizePhone
    'shipping.phone': normalizePhone
  }
})
```
Needs investigation and trial and error iteration...

If you do decide to act on this, please use the [Examples repo](https://github.com/erikras/redux-form/tree/gh-pages) as your test bed. Maybe the [Deep Form example](http://erikras.github.io/redux-form/#/examples/deep) could use a normalizer to keep the city in title case and format the phone number array...

**Comment by ekosz** [comment]
Just ran into this. Normalizing deep fields would be great!

**Comment by tokenvolt** [comment]
Same here, I'm currently doing some workarounds in the app, which could be avoided by normalizing.

**Comment by bradennapier** [comment]
Yes - deep field normalization definitely is needed!!! Adding users in this way and can't normalize any of the data that is created by it.

I second this method: 
     reducer.normalize({
       contactForm: {
         'billing.phone': normalizePhone
         'shipping.phone': normalizePhone
       }
     })

**Comment by JohnPhoto** [comment]
I too need the deep field normalization. My usecase is that of taking the value of 2 fields and merge them with some extra logic into a third (using allValues).

**Comment by trashgenerator** [comment]
In case somebody don't know about redux-form plugins: you can use plugins system for normalizing deep fields right now
http://erikras.github.io/redux-form/#/api/reducer/plugin?_k=iob9yf

**Comment by JohnPhoto** [comment]
Will most certainly look into this, thank you for ProTip.

**Comment by dozoisch** [comment]
That would be really awesome! I ran into that also today!

**Comment by erikras** [comment]
@jfairbank has fixed this with PR #637. Published as [`v4.1.12`](https://github.com/erikras/redux-form/releases/tag/v4.1.12).

**Comment by CptCompiler** [comment]
Hi,
I have the following fields:
`form: {
      'timeSlots[].startsAt': normalizeStartsAt,
      'timeSlots[].endsAt': normalizeEndsAt
    }`

When I normalize startsAt I want to know what value endsAt has. Is there any way to get the index of the current timeSlot? So that I can get the value with allValues.timeSlots[index].endsAt

What I want to do is to make sure that startsAt is always smaller than endsAt. Is there a better solution to do this than with normalize?

Thanks,
Andres
